# Homework 1

## 1. Install Python (if you don’t have it already), and install NLTK.

In [55]:
from nltk.book import *
import re

## 2. Follow the instructions in chapter 1 of Bird-Klein for implementing a “lexical diversity” scoring routine.

## 3.	Go to http://www.gutenberg.org/wiki/Children%27s_Instructional_Books_(Bookshelf), and obtain three texts (of different grade levels) from the “Graded Readers” section. Report the lexical diversity score of each. Explain whether the result was surprising.

In [50]:
start_line = '\*\*\* START OF THIS PROJECT GUTENBERG EBOOK.*\*\*\*'
end_line = '\*\*\* END OF THIS PROJECT GUTENBERG EBOOK.*\*\*\*'

In [56]:
def lexical_diversity(text):
    if text is str:
        text = split_text(text)
    return 100 * len(set(text)) / len(text)

def vocab_score(text):
    if text is str:
        text = split_text(text)
    return len(set(text))

def split_text(text):
    return re.split("[^a-z']+")

def remove_meta(text):
    return re.split(end_line, re.split(start_line, text)[1])[0]

In [44]:
grade2 = open('data/beacon_second_reader.txt').read()
grade4 = open('data/mcguffys_fourth_reader.txt').read()
grade5 = open('data/mcguffys_fifth_reader.txt').read()
gradehs = open('data/ontario_highschool_reader.txt').read()

In [57]:
print lexical_diversity(remove_meta(grade2))

0.06811473548777719

In [62]:
print lexical_diversity(remove_meta(grade4))

0.0230104933394325

In [63]:
print lexical_diversity(remove_meta(grade5))

0.010439217655202584

In [64]:
print lexical_diversity(remove_meta(gradehs))

0.026381771234729214

###	The following is the lexical diversity scores for each of the books I checked.
* The Bacon Second Reader

>* 	lexical diversity = .06811 % 

*	Mcguffy’s Forth Eclectic Reader

>*	lexical diversity = .02301 %

*	Mcguffy’s Fifth Eclectic Reader

>*	lexical diversity = .01043

*	The Ontario High School Reader

>*	lexical diversity = .02638

*	The surprising thig here is that the book with the highest lexical diversity is the second grade reader (or the least advanced book).  This may be because the least advanced book is simply shorter then the other books, and therefore has a less significant lexical diversity.  In fact, in the books listed, each subsequent grade, with the exception of high school, has a lower lexical diversity then the grades below it.

In [58]:
print vocab_score(remove_meta(grade2))

81

In [35]:
print vocab_score(remove_meta(grade4))

87

In [59]:
print vocab_score(remove_meta(grade5))

84

In [65]:
print vocab_score(remove_meta(gradehs))

127

## 4.	Also compare the vocabulary size of the same three texts. Explain whether the result was surprising.

Vocabulary counts for all books
* The Bacon Second Reader
>* Unique word count = 81 
* Mcguffy’s Forth Eclectic Reader
>* Unique word count = 87
* Mcguffy’s Fifth Eclectic Reader
>* Unique word count = 84
* The Ontario High School Reader
>* Unique word count = 127

-	In the case of vocabulary counts, the result is more in line with expectations.   This shows that higher reading level books have a larger vocabulary.  The only exception here is the fourth grade to fifth grade transition; where fifth grade has slightly less overall unique words then fourth.

## 5.	Write a paragraph arguing whether vocabulary size and lexical diversity in combination could be a better measure of text difficulty (or reading level) than either measure is by itself.

-	Based on the data collected in this assignment, Vocabulary size seems to be a better measure of reading level then lexical diversity, but even with this measure, there is still an example that diverges from the trend.  Perhaps the best method of measuring text reading difficulty would not be to measure the overall word count or lexical diversity, but to measure the frequency of word appearance in relation to how often the words appear in all English language.  It also seems probable that it is not easy to measure reading difficulty by measuring a single variable.  It would have to be placed within a larger context.  For example, if one book has a very large vocabulary of commonly used words; while another book contains more ecelectic language, but contains overall less vocabulary, the later book would likely be the more difficult one to read.